compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,268 @@
+use crate::transform::MirPass;
+use rustc_data_structures::stable_map::FxHashMap;
+use rustc_middle::mir::interpret::Allocation;
+use rustc_middle::mir::*;
+use rustc_middle::ty::util::IntTypeExt;
+use rustc_middle::ty::{self, AdtDef, Const, ParamEnv, Ty, TyCtxt};
+use rustc_target::abi::{HasDataLayout, Size, TagEncoding, Variants};
+use std::array::IntoIter;
+
+/// A pass that seeks to optimize unnecessary moves of large enum types, if there is a large
+/// enough discrepanc between them
+pub struct EnumSizeOpt {
+    pub(crate) discrepancy: u64,
+}
+
+impl<'tcx> MirPass<'tcx> for EnumSizeOpt {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        self.optim(tcx, body);
+    }
+}
+
+impl EnumSizeOpt {
+    fn candidate<'tcx>(
+        &self,
+        tcx: TyCtxt<'tcx>,
+        param_env: ParamEnv<'tcx>,
+        ty: Ty<'tcx>,
+        alloc_cache: &mut FxHashMap<Ty<'tcx>, &'tcx Allocation<(), ()>>,
+    ) -> Option<(&'tcx AdtDef, usize, &'tcx Allocation<(), ()>)> {
+        let adt_def = match ty.kind() {
+            ty::Adt(adt_def, _substs) if adt_def.is_enum() => adt_def,
+            _ => return None,
+        };
+        let layout = if let Ok(layout) = tcx.layout_of(param_env.and(ty)) {
+            layout
+        } else {
+            return None;
+        };
+        let variants = match &layout.variants {
+            Variants::Single { .. } => return None,
+            Variants::Multiple { tag_encoding, .. }
+                if matches!(tag_encoding, TagEncoding::Niche { .. }) =>
+            {
+                return None;
+            }
+            Variants::Multiple { variants, .. } if variants.len() <= 1 => return None,
+            Variants::Multiple { variants, .. } => variants,
+        };
+        let min = variants.iter().map(|v| v.size).min().unwrap();
+        let max = variants.iter().map(|v| v.size).max().unwrap();
+        if max.bytes() - min.bytes() < self.discrepancy {
+            return None;
+        }
+
+        let num_discrs = adt_def.discriminants(tcx).count();
+        if variants.iter_enumerated().any(|(var_idx, _)| {
+            adt_def.discriminant_for_variant(tcx, var_idx).val as usize >= num_discrs
+        }) {
+            return None;
+        }
+        if let Some(alloc) = alloc_cache.get(ty) {
+            return Some((adt_def, num_discrs, alloc));
+        }
+
+        let data_layout = tcx.data_layout();
+        let ptr_sized_int = data_layout.ptr_sized_integer();
+        let target_bytes = ptr_sized_int.size().bytes() as usize;
+        let mut data = vec![0; target_bytes * num_discrs];
```

Comment:
It might be slightly simpler to start with an empty vec (with expected capacity), and extend it with byte slices, avoiding all offset calculations in the process.

Reply:
I assume that discriminants may be out of order, but if that assumption is wrong I agree your approach may be cleaner. Unless it's always the case that discriminants are in order I'll leave as is